explorer/src/components/account/AnchorAccountCard.tsx

```diff
@@ -1,63 +1,62 @@
-    account.details?.owner.toString() ?? "",
+  const anchorProgram = useAnchorProgram(
+    account.details?.owner.toString() || "",
     url
   );
+  const rawData = account?.details?.rawData;
+  const programName = getProgramName(anchorProgram) || "Unknown Program";
 
-  const { foundAccountLayoutName, decodedAnchorAccountData } = useMemo(() => {
-    let foundAccountLayoutName: string | undefined;
-    let decodedAnchorAccountData: { [key: string]: any } | undefined;
-    if (program && account.details && account.details.rawData) {
-      const accountBuffer = account.details.rawData;
-      const discriminator = accountBuffer.slice(0, 8);
-
-      // Iterate all the structs, see if any of the name-hashes match
-      Object.keys(program.account).forEach((accountType) => {
-        const layoutName = capitalizeFirstLetter(accountType);
-        const discriminatorToCheck =
-          BorshAccountsCoder.accountDiscriminator(layoutName);
-
-        if (discriminatorToCheck.equals(discriminator)) {
-          foundAccountLayoutName = layoutName;
-          const accountDecoder = program.account[accountType];
-          decodedAnchorAccountData = accountDecoder.coder.accounts.decode(
-            layoutName,
-            accountBuffer
-          );
-        }
-      });
+  const { decodedAccountData, accountDef } = useMemo(() => {
+    let decodedAccountData: any | null = null;
+    let accountDef: IdlTypeDef | undefined = undefined;
+    if (anchorProgram && rawData) {
+      const coder = new BorshAccountsCoder(anchorProgram.idl);
+      const accountDefTmp = anchorProgram.idl.accounts?.find(
+        (accountType: any) =>
+          (rawData as Buffer)
+            .slice(0, 8)
+            .equals(BorshAccountsCoder.accountDiscriminator(accountType.name))
+      );
+      if (accountDefTmp) {
+        accountDef = accountDefTmp;
+        decodedAccountData = coder.decode(accountDef.name, rawData);
+      }
     }
-    return { foundAccountLayoutName, decodedAnchorAccountData };
-  }, [program, account.details]);
 
-  if (!foundAccountLayoutName || !decodedAnchorAccountData) {
+    return {
+      decodedAccountData,
+      accountDef,
+    };
+  }, [anchorProgram, rawData]);
+
+  if (lamports === undefined) return null;
+  if (!anchorProgram) return <ErrorCard text="No Anchor IDL found" />;
+  if (!decodedAccountData || !accountDef) {
     return (
-      <ErrorCard text="Failed to decode account data according to its public anchor interface" />
+      <ErrorCard text="Failed to decode account data according to the public Anchor interface" />
     );
   }
 
   return (
-    <>
+    <div>
       <div className="card">
         <div className="card-header">
           <div className="row align-items-center">
             <div className="col">
-              <h3 className="card-header-title">{foundAccountLayoutName}</h3>
+              <h3 className="card-header-title">
+                {programName} {">"} {accountDef.name}
```

Comment:
For example, in /tx/bk9u61zRcbuXTHsPJVSqSCunRqVNCcziXbNAqs4rnCM1UPD1o6jFzgsiKZjA4sSajUM8dM1FRdEm9tRiJ251nvM

i.e. with the `<ProgramName ... link>` change, the program name "Drift Client" can be underlined in the instruction card title, without repetitively underlining it in the Program Logs.

Reply:
cc @steveluscher might be worth punting this into a follow up PR that I can be responsible for. These changes are probably worth taking ASAP to reduce throwing Errors client side

Reply:
That's a good point. It made me realize that I deleted the `Program` row for the instruction details. I added it back and used a slightly modified `Address` to get this:

![image](https://user-images.githubusercontent.com/95379755/163092325-5b059ac2-645e-43e4-9675-14aeaf1d8433.png)

In itself it's ok I think, but imo the best would be to integrate decoding mechanisms like Anchor into the `Address` component itself. After all it is where the name fetching logic resides atm (raw address, static naming tables, metadata).

I could modify the this component in a similar fashion as `ProgramName`, using `Suspense` to abstract the new async logic and be transparent to existing components using it. The link with IDL program name would then naturally appear in the account header section, and we can leave the `Anchor Account Data` card title as simply: "{program name}: {account type}", to match what is done in instruction details.

Wdyt?

Reply:
I think the Anchor Program name should have a different UI than normal Addresses / Program Names.

Worst case I'm imagining right now is that someone deploys Anchor contracts with names that mimic SPL program names, and then convinces people to interact with their contracts. They can send explorer links & it will look indistinguishable from the original SPL programs.

I'd prefer just having something like __Exchange__(*): My Account Name, with an on-hover mechanic which says that this the name given to it by IDL, rather than a name that we have manually confirmed.

Right now I'm comfortable with just using `{programName}: {account type}` as in the instruction details.

Reply:
That's an interesting point. I guess I am coming at it from the POV that a user would have to click on the `Address` to verify that the program is the one they expect, if they have a Pubkey in mind. Although the only way then would be to look at the URL bar. I'll let you figure out the cleanest way!